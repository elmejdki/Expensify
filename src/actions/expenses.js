import database from '../firebase/firebase';

export const addExpense = (expense = {}) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    const {
      description = "",
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
      dispatch(addExpense({
        id: snapshot.key,
        ...expense
      }));
    });
  }
}

export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = (id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id));
    });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
}

export const startSetExpenses = () => {
  return (disptach, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];

      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      disptach(setExpenses(expenses));
    });
  };
};
