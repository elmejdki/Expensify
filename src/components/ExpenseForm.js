import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  state = {
    description: this.props.description || '',
    note: this.props.note || '',
    amount: this.props.amount ? this.props.amount / 100 + '' : '',
    createdAt: this.props.createdAt ? moment(this.props.createdAt) : moment(),
    calendarFocused: false,
    state: ''
  }

  onDescriptionChange = (event) => {
    this.setState(() => ({
      description: event.target.value
    }));
  }

  onNoteChange = (event) => {
    this.setState(() => ({
      note: event.target.value
    }))
  }

  onAmountChange = (event) => {
    const amount = event.target.value;
    if(!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      // set error state equal to 'Please provide description and amount'
      this.setState(() => ({
        error: 'Please provide description and amount'
      }));
    }else {
      this.setState(() => ({
        error: ''
      }));
      const { description, note, amount, createdAt } = this.state;
      this.props.onSubmit({
        description,
        note,
        createdAt: createdAt.valueOf(),
        amount: parseInt(Number(amount)*100)  
      });
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        { this.state.error && <p className="form__error">{this.state.error}</p>}
        <input 
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          value={this.state.note}
          className="textarea"
          onChange={this.onNoteChange}
          placeholder="Add a note for your expense (optional)"
        >
        </textarea>
        <div>
          <button className="button">
            {
              this.props.isEdit ? 'Edit Expense' : 'Add Expense'
            }
          </button>
        </div>
      </form>
    );
  }
}
