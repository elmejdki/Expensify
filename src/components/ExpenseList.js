import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No Expenses</p>
      ) : (
        props.expenses.map(expense => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;
