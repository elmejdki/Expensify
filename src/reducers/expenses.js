const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState,
  action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state, action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter((expense) => expense.id !== action.id );
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if(expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        }else {
          return expense;
        }
      });
    default:
      return state;
  };
}

export default expensesReducer;