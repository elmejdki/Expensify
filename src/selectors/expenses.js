const getVisibleExpenses = (expenses, { text, startDate, endDate,sortBy }) => {
  return expenses.filter(
    (expense) => {
      let {description, note} = expense;
      description = description.toLowerCase();
      note = note.toLowerCase();
      text = text.toLowerCase();

      return (
        description.includes(text) || note.includes(text)
      ) && (
        (
          typeof startDate !== 'number' || expense.createdAt >= startDate
        ) && (
          typeof endDate !== 'number' || expense.createdAt <= endDate
        )
      )
  }).sort((expense_a, expense_b) => 
    sortBy === 'amount' ?
      expense_b.amount - expense_a.amount :
      expense_b.createdAt - expense_a.createdAt
  );
}

export default getVisibleExpenses;