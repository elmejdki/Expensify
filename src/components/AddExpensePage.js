import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        props.addExpense(expense);
        props.history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = { addExpense };

export default connect(null, mapDispatchToProps)(AddExpensePage);