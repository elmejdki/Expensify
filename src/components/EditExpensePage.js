import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export const EditExpensePage = (props) => {
  const id = props.match.params.id;

  return (
    <div>
      <ExpenseForm
        {...props.expense}
        onSubmit={(expense) => {
          props.editExpense(id, expense);
          props.history.push('/');
        }}
        isEdit={true}
      />
      <button
        onClick={() => {
          props.startRemoveExpense(id);
          props.history.push('/');
        }}
      >
        Remove
      </button>
    </div>
  )
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  };
}

const mapDispatchToProps = { editExpense, startRemoveExpense };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);