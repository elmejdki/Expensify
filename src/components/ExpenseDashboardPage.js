import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;