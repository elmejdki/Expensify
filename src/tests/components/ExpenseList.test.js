import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

const renderer = new ShallowRenderer();

it('should render ExpenseList with expenses', () => {
  renderer.render(<ExpenseList expenses={expenses} />);

  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should render ExpenseList with empty message', () => {
  renderer.render(<ExpenseList expenses={[]} />);

  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
