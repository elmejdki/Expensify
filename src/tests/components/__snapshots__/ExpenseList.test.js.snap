// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    note=""
  />
</div>
`;
